Make tests spec-compliant and pass on Lotus

arajasek opened the pull request:
**PR #113 should be merged first (or closed after this PR is merged)**

- provide internalCallSeqNum when calculating expected actor addresses
- use {idaddress, outer address} structs in expected message receipts
- use pubkey address, not ID address, to calculate expected message receipts
- expect txnId, not empty return, for multisig propose message

frrist commented on suites/message/create_actor.go:
```diff
@@ -109,15 +109,15 @@ func TestInitActorSequentialIDAddressCreate(t *testing.T, factory state.Factorie
-	firstInitRet := td.ComputeInitActorExecReturn(senderID, 0, firstPaychAddr)
-	secondInitRet := td.ComputeInitActorExecReturn(senderID, 1, secondPaychAddr)
+	firstInitRet := td.ComputeInitActorExecReturn(sender, 1, firstPaychAddr)
+	secondInitRet := td.ComputeInitActorExecReturn(sender, 2, secondPaychAddr)
```
I think this should continue to pass the ID based address here since it represents the runtimes immediate caller and is always expected to be an ID-address: https://filecoin-project.github.io/specs/#vmruntime-interface

arajasek replied:
Thanks! I thought the spec was ambiguous here and picked the pubkey address (since that's what Lotus was doing), but it is very much not ambiguous. I've changed Lotus to do the correct thing, and undone my changes here :)

frrist commented on suites/message/create_actor.go:
```diff
@@ -116,8 +116,8 @@ func TestInitActorSequentialIDAddressCreate(t *testing.T, factory state.Factorie
-	firstInitRet := td.ComputeInitActorExecReturn(senderID, 0, firstPaychAddr)
-	secondInitRet := td.ComputeInitActorExecReturn(senderID, 1, secondPaychAddr)
+	firstInitRet := td.ComputeInitActorExecReturn(senderID, 1, firstPaychAddr)
+	secondInitRet := td.ComputeInitActorExecReturn(senderID, 2, secondPaychAddr)
```
Sorry I didn't notice this change first review. I don't think this is right, `0` and `1` are the CallSeqNum of the from (`sender`) Actor. Actors are initialized with `CallSeqNum: 0` and `sender` shouldn't have had their `CallSeqNum` incremented at this point, unless I am missing something?

arajasek replied:
If I'm understanding the spec correctly, they *should* have it incremented by this point. I'm basing that off the `ApplyMessage` function here: https://filecoin-project.github.io/specs/#vminterpreter-implementation

AFAIC tell, the rule is that you increment the nonce as soon as the basic validity checks have passed, and *before* any actual actor code is invoked.

Whether that *should* be the rule is ofc up for debate, but rn it seems that the spec says that the `CallSeqNum`s should already have been incremented.

anorth replied:
That is what the spec code says, indeed. It wasn't consciously that way. [Discussion](https://filecoinproject.slack.com/archives/CHMNDCK9P/p1583875804246100)

arajasek committed:
Fix failing tests

- provide internalCallSeqNum when calculating expected actor addresses
- use {idaddress, outer address} structs in expected message receipts
- use pubkey address, not ID address, to calculate expected message receipts
- expect txnId, not empty return, for multisig propose message

arajasek commented on suites/tipset/miner_flow.go:
```diff
@@ -71,15 +71,15 @@ func TestMinerCreateProveCommitAndMissPoStChallengeWindow(t *testing.T, factory
 				chain.Nonce(0), chain.Value(collateral),
 			),
-			types.MessageReceipt{ExitCode: exitcode_spec.Ok, ReturnValue: []byte{}, GasUsed: big_spec.Zero()},
+			types.MessageReceipt{ExitCode: exitcode_spec.Ok, ReturnValue: nil, GasUsed: big_spec.Zero()},
```
Can someone confirm that this is correct? It looks like nil is what the specs actor returns, but there might be something I'm missing.

frrist replied:
Yes `nil` is the correct value. 

Note that lotus currently skips validation of the message receipt return field: https://github.com/filecoin-project/lotus/blob/testnet/3/chain/validation/factories.go#L44

arajasek commented on suites/tipset/miner_flow.go:
```diff
@@ -42,7 +42,7 @@ func TestMinerCreateProveCommitAndMissPoStChallengeWindow(t *testing.T, factory
 	minerWorker, minerWorkerID := td.NewAccountActor(address.BLS, abi_spec.NewTokenAmount(1_000_000_000))
 	// The address of the miner actor
 	minerActorID := utils.NewIDAddr(t, utils.IdFromAddress(minerWorkerID)+1)
-	createMinerRet := td.ComputeInitActorExecReturn(builtin_spec.StoragePowerActorAddr, 0, minerActorID)
```
pretty sure this was just wrong, but i might be missing something.

frrist replied:
Since a storage miner actor is [created by](https://github.com/filecoin-project/specs-actors/blob/3f5510bf61303d1f629f8fce26ebe3194c911f93/actors/builtin/miner/miner_actor.go#L69-L70) the storage power actor I believe the StoragePowerActorAddr is the correct parameter here.

frrist commented on suites/message/create_actor.go:
```diff
@@ -109,15 +109,15 @@ func TestInitActorSequentialIDAddressCreate(t *testing.T, factory state.Factorie
 
-	firstInitRet := td.ComputeInitActorExecReturn(senderID, 0, firstPaychAddr)
-	secondInitRet := td.ComputeInitActorExecReturn(senderID, 1, secondPaychAddr)
+	firstInitRet := td.ComputeInitActorExecReturn(sender, 0, 0, firstPaychAddr)
+	secondInitRet := td.ComputeInitActorExecReturn(sender, 1, 0, secondPaychAddr)
```
I believe the `senderID` address should be passed here since `ApplyMessage` will [resolve the sender](https://filecoin-project.github.io/specs/#vminterpreter-implementation) address through the InitActor's map before performing any operations.

Currently go-filecoin performs the operations as described in the spec here: https://github.com/filecoin-project/go-filecoin/blob/master/internal/pkg/vm/internal/vmcontext/vmcontext.go#L360

From a quick skim of the lotus code I don't believe this has been instrumented yet: https://github.com/filecoin-project/lotus/blob/testnet/3/chain/vm/vm.go#L252

frrist commented on suites/message/message_application.go:
```diff
@@ -90,18 +91,21 @@ func TestMessageApplicationEdgecases(t *testing.T, factory state.Factories) {
 
 		// the _expected_ address of the payment channel
 		paychAddr := utils.NewIDAddr(t, utils.IdFromAddress(receiverID)+1)
-		createRet := td.ComputeInitActorExecReturn(senderID, 0, paychAddr)
```
:arrow_up: ditto the comment about senderID address here

frrist commented on suites/message/message_application.go:
```diff
@@ -116,5 +120,11 @@ func TestMessageApplicationEdgecases(t *testing.T, factory state.Factories) {
+		gasReward := big.NewInt(td.MessageProducer.DefaultGasLimit())
+		gasReward = gasReward.Mul(gasReward, td.MessageProducer.DefaultGasPrice().Int)
+		senderBalance.Sub(senderBalance.Int, gasReward)
+		td.AssertBalance(sender, senderBalance)
```
Unsure this is the correct behaviour. The [ApplyMessage](https://filecoin-project.github.io/specs/#vminterpreter-implementation) section of the spec indicates that unused gas is to be refunded to the caller iff the sender address is valid (which it is here).

anorth replied:
This change is invalid. Only the actually consumed gas should be spent.

If you believe there is disagreement about this, please remove this change so that this PR can otherwise land, and open an issue for discussion in specs.

frrist commented on suites/message/multisig.go:
```diff
@@ -38,7 +38,7 @@ func TestMultiSigActor(t *testing.T, factory state.Factories) {
 		// expected address of the actor
 		multisigAddr := utils.NewIDAddr(t, 1+utils.IdFromAddress(aliceId))
 
-		createRet := td.ComputeInitActorExecReturn(aliceId, 0, multisigAddr)
```
:arrow_up: dittio comments about senderID address here.

frrist commented on suites/message/multisig.go:
```diff
@@ -403,7 +406,7 @@ func TestMultiSigActor(t *testing.T, factory state.Factories) {
 
 		multisigAddr := utils.NewIDAddr(t, utils.IdFromAddress(chuckId)+1)
-
+		createRet := td.ComputeInitActorExecReturn(alice, 0, 0, multisigAddr)
```
senderID here as well.

frrist reviewed (changes requested):
Thanks for moving this forward, I think we need to have some more discussion here before this is ready to go in.

frrist reviewed (approved):
After reading https://filecoinproject.slack.com/archives/CHMNDCK9P/p1583985982257700 I understand that these changes are to be considered valid, the spec just doesn't reflect that yet. Let's hold on merging this for a little bit until some context landes in github.com/filecoin-project/specs. 

Thanks @arajasek for driving this forward.

arajasek commented:
@anorth I've rolled back the test on gas cost for reverts. I think Lotus might not be spec-compliant in its behaviour there, I'll look into it and add tests here later if necessary.
